File: src/test/java/com/barracuda/eshop/customer/ExceptionTranslatorTest.java
package com.barracuda.eshop.customer;

import com.barracuda.eshop.customer.exception.DuplicateEmailCustomerException;
import com.barracuda.eshop.customer.exception.ExceptionTranslator;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.dao.DuplicateKeyException;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class ExceptionTranslatorTest {

    private final ExceptionTranslator exceptionTranslator = new ExceptionTranslator();

    @Test
    void shouldTranslateDuplicateKeyExceptionToDuplicateCustomerEmailExceptionIfTheExceptionMessageContainsEmailConstraintViolation() {

        assertThatThrownBy(() -> exceptionTranslator.translate(new DuplicateKeyException("email_unique_constraint"))).isInstanceOf(DuplicateEmailCustomerException.class);

    }

    @Test
    void shouldNotTranslateRethrowTheProvidedExceptionIfTranslationIsNotPossible() {
        var exception = new DuplicateKeyException("UNKNOWN_ERROR");
        assertThatThrownBy(() -> exceptionTranslator.translate(exception)).isEqualTo(exception);
    }
}

File: src/test/kotlin/com/barracuda/eshop/customer/exception/ExceptionTranslatorTest.kt
package com.barracuda.eshop.customer.exception

import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.springframework.dao.DuplicateKeyException

class ExceptionTranslatorTest {
    private val exceptionTranslator = ExceptionTranslator()

    @Test
    fun `should translate DuplicateKeyException to DuplicateCustomerEmailException if the exception message contains unique email constraint violation`() {
        assertThatThrownBy {
            exceptionTranslator.translate(DuplicateKeyException("email_unique_constraint"))
        }.isInstanceOf(DuplicateEmailCustomerException::class.java)
    }

    @Test
    fun `should not translate and rethrow the provided exception if the exception is not recognized`() {
        val exception = DuplicateKeyException("UNKNOWN_ERROR")
        assertThatThrownBy { exceptionTranslator.translate(exception) }.isEqualTo(exception)
    }
}
